Calling an internal type constructor in a public operation generates invalid C#

## Describe the bug
Calling an internal type constructor in a public operation generates a C# class for the operation that uses the internal type as the type of a protected field, which is not legal.

## To Reproduce
```
internal newtype MyInt = Int;

operation HelloQ() : Unit {
    let x = MyInt(3);
}
```

## Expected behavior
The generated C# code should compile.

## Actual behavior
> obj\qsharp\src\Program.g.cs(49,43): error CS0053: Inconsistent access: Property type "ICallable<long, MyInt>" is less accessible than property "HelloQ.MyInt". [C:\Users\samarsha\Desktop\QsSandbox\QsSandbox.csproj]

## System information
- OS: Windows
- .NET Core 3.1.200
- Q# SDK 0.11.2003.2505-beta

**Additional context**
The problem property is:

```
public partial class HelloQ : Operation<QVoid, QVoid>, ICallable
{
    // ...

    protected ICallable<Int64, MyInt> MyInt
    {
        get;
        set;
    }

    // ...
}
```

--- a/src/Simulation/CsharpGeneration.Tests/Circuits/CodegenTests.qs
+++ b/src/Simulation/CsharpGeneration.Tests/Circuits/CodegenTests.qs
@@ -1309,11 +1309,20 @@ namespace Microsoft.Quantum.Compiler.Generics {
 
     // Access Modifiers
 
-    internal function InternalFunction () : Unit {
-    }
+    internal function EmptyInternalFunction () : Unit { }
 
-    internal operation InternalOperation () : Unit {
-    }
+    internal operation EmptyInternalOperation () : Unit { }
 
     internal newtype InternalType = Unit;
+
+    internal operation MakeInternalType () : InternalType {
+        return InternalType();
+    }
+
+    operation UseInternalCallables () : Unit {
+        EmptyInternalFunction();
+        EmptyInternalOperation();
+        let x = InternalType();
+        let y = MakeInternalType();
+    }
 }

--- a/src/Simulation/CsharpGeneration.Tests/SimulationCodeTests.fs
+++ b/src/Simulation/CsharpGeneration.Tests/SimulationCodeTests.fs
@@ -194,8 +194,9 @@ namespace N1
     let bitOperations                           = findCallable @"bitOperations"
     let testLengthDependency                    = findCallable @"testLengthDependency"
     let UpdateUdtItems                          = findCallable @"UpdateUdtItems"
-    let internalFunction                        = findCallable @"InternalFunction"
-    let internalOperation                       = findCallable @"InternalOperation"
+    let emptyInternalFunction                   = findCallable @"EmptyInternalFunction"
+    let emptyInternalOperation                  = findCallable @"EmptyInternalOperation"
+    let useInternalCallables                    = findCallable @"UseInternalCallables"
 
     let udt_args0                               = findUdt @"udt_args0"
     let udt_args1                               = findUdt @"udt_args1"
@@ -2565,18 +2566,18 @@ namespace N1
     [<Fact>]
     let ``buildOperationClass - access modifiers`` () =
         """
-[SourceLocation("%%%", OperationFunctor.Body, 1312, 1315)]
-internal partial class InternalFunction : Function<QVoid, QVoid>, ICallable
+[SourceLocation("%%%", OperationFunctor.Body, 1312, 1314)]
+internal partial class EmptyInternalFunction : Function<QVoid, QVoid>, ICallable
 {
-    public InternalFunction(IOperationFactory m) : base(m)
+    public EmptyInternalFunction(IOperationFactory m) : base(m)
     {
     }
 
-    String ICallable.Name => "InternalFunction";
+    String ICallable.Name => "EmptyInternalFunction";
 
-    String ICallable.FullName => "Microsoft.Quantum.Compiler.Generics.InternalFunction";
+    String ICallable.FullName => "Microsoft.Quantum.Compiler.Generics.EmptyInternalFunction";
 
-    public static OperationInfo<QVoid, QVoid> Info => new OperationInfo<QVoid, QVoid>(typeof(InternalFunction));
+    public static OperationInfo<QVoid, QVoid> Info => new OperationInfo<QVoid, QVoid>(typeof(EmptyInternalFunction));
 
     public override Func<QVoid, QVoid> Body => (__in__) =>
     {
@@ -2594,25 +2595,25 @@ internal partial class InternalFunction : Function<QVoid, QVoid>, ICallable
 
     public static System.Threading.Tasks.Task<QVoid> Run(IOperationFactory __m__)
     {
-        return __m__.Run<InternalFunction, QVoid, QVoid>(QVoid.Instance);
+        return __m__.Run<EmptyInternalFunction, QVoid, QVoid>(QVoid.Instance);
     }
 }
 """
-        |> testOneClass internalFunction
+        |> testOneClass emptyInternalFunction
 
         """
-[SourceLocation("%%%", OperationFunctor.Body, 1315, 1318)]
-internal partial class InternalOperation : Operation<QVoid, QVoid>, ICallable
+[SourceLocation("%%%", OperationFunctor.Body, 1314, 1316)]
+internal partial class EmptyInternalOperation : Operation<QVoid, QVoid>, ICallable
 {
-    public InternalOperation(IOperationFactory m) : base(m)
+    public EmptyInternalOperation(IOperationFactory m) : base(m)
     {
     }
 
-    String ICallable.Name => "InternalOperation";
+    String ICallable.Name => "EmptyInternalOperation";
 
-    String ICallable.FullName => "Microsoft.Quantum.Compiler.Generics.InternalOperation";
+    String ICallable.FullName => "Microsoft.Quantum.Compiler.Generics.EmptyInternalOperation";
 
-    public static OperationInfo<QVoid, QVoid> Info => new OperationInfo<QVoid, QVoid>(typeof(InternalOperation));
+    public static OperationInfo<QVoid, QVoid> Info => new OperationInfo<QVoid, QVoid>(typeof(EmptyInternalOperation));
 
     public override Func<QVoid, QVoid> Body => (__in__) =>
     {
@@ -2630,11 +2631,11 @@ internal partial class InternalOperation : Operation<QVoid, QVoid>, ICallable
 
     public static System.Threading.Tasks.Task<QVoid> Run(IOperationFactory __m__)
     {
-        return __m__.Run<InternalOperation, QVoid, QVoid>(QVoid.Instance);
+        return __m__.Run<EmptyInternalOperation, QVoid, QVoid>(QVoid.Instance);
     }
 }
 """
-        |> testOneClass internalOperation
+        |> testOneClass emptyInternalOperation
 
 
     [<Fact>]
@@ -2696,7 +2697,26 @@ internal partial class InternalOperation : Operation<QVoid, QVoid>, ICallable
             |> List.map formatSyntaxTree
             
         List.zip (expected |> List.map clearFormatting) (actual  |> List.map clearFormatting) |> List.iter Assert.Equal
-        
+
+    [<Fact>]
+    let ``buildOpsProperties - internal callables`` () =
+        let property = sprintf "private protected %s %s { get; set; }"
+        let expected =
+            [
+                property "ICallable<QVoid, QVoid>" "EmptyInternalFunction"
+                property "ICallable<QVoid, QVoid>" "EmptyInternalOperation"
+                property "ICallable<QVoid, InternalType>" "InternalType"
+                property "ICallable<QVoid, InternalType>" "MakeInternalType"
+            ]
+        let op = snd useInternalCallables
+        let actual =
+            op
+            |> operationDependencies
+            |> depsByName
+            |> buildOpsProperties (createTestContext op)
+            |> List.map formatSyntaxTree
+        List.zip (List.map clearFormatting expected) (List.map clearFormatting actual)
+        |> List.iter Assert.Equal
 
     [<Fact>]
     let ``buildOperationClass - concrete functions`` () = 

--- a/src/Simulation/CsharpGeneration/SimulationCode.fs
+++ b/src/Simulation/CsharpGeneration/SimulationCode.fs
@@ -904,15 +904,15 @@ module SimulationCode =
     /// a Property that returns an instance of the operation by calling the
     /// IOperationFactory
     let buildOpsProperties context (operations : QsQualifiedName list): MemberDeclarationSyntax list =
-        let getAccessModifier qualifiedName =
+        let getCallableAccessModifier qualifiedName =
             match context.allCallables.TryGetValue qualifiedName with
             | true, callable -> Some callable.Modifiers.Access
             | false, _ -> None
 
-        let getModifiers qualifiedName =
+        let getPropertyModifiers qualifiedName =
             // Use the right accessibility for the property depending on the accessibility of the callable.
             // Note: In C#, "private protected" is the intersection of protected and internal.
-            match getAccessModifier qualifiedName |> Option.defaultValue DefaultAccess with
+            match getCallableAccessModifier qualifiedName |> Option.defaultValue DefaultAccess with
             | DefaultAccess -> [ ``protected`` ]
             | Internal -> [ ``private``; ``protected`` ]
 
@@ -921,7 +921,7 @@ module SimulationCode =
             /// protected opType opName { get; }
             let signature = roslynCallableTypeName context qualifiedName
             let name = getOpName context qualifiedName
-            let modifiers = getModifiers qualifiedName
+            let modifiers = getPropertyModifiers qualifiedName
             ``prop`` signature name modifiers
             :> MemberDeclarationSyntax
 

--- a/src/Simulation/CsharpGeneration/SimulationCode.fs
+++ b/src/Simulation/CsharpGeneration/SimulationCode.fs
@@ -904,12 +904,18 @@ module SimulationCode =
     /// a Property that returns an instance of the operation by calling the
     /// IOperationFactory
     let buildOpsProperties context (operations : QsQualifiedName list): MemberDeclarationSyntax list =
+        let getAccessModifier qualifiedName =
+            match context.allCallables.TryGetValue qualifiedName with
+            | true, callable -> Some callable.Modifiers.Access
+            | false, _ -> None
+
         let getModifiers qualifiedName =
-            // Use the right accessibility for the property depending on the accessibility of the callable. Note: In C#,
-            // "private protected" is the intersection of protected and internal.
-            match context.allCallables.[qualifiedName].Modifiers.Access with
+            // Use the right accessibility for the property depending on the accessibility of the callable.
+            // Note: In C#, "private protected" is the intersection of protected and internal.
+            match getAccessModifier qualifiedName |> Option.defaultValue DefaultAccess with
             | DefaultAccess -> [ ``protected`` ]
             | Internal -> [ ``private``; ``protected`` ]
+
         let buildOne qualifiedName =
             /// eg:
             /// protected opType opName { get; }
@@ -918,6 +924,7 @@ module SimulationCode =
             let modifiers = getModifiers qualifiedName
             ``prop`` signature name modifiers
             :> MemberDeclarationSyntax
+
         operations
         |> List.map buildOne
 

--- a/src/Simulation/CsharpGeneration/SimulationCode.fs
+++ b/src/Simulation/CsharpGeneration/SimulationCode.fs
@@ -899,17 +899,24 @@ module SimulationCode =
             [ ``public`` ]
             ``{`` [] ``}``
         :> MemberDeclarationSyntax     
-       
+
     /// For each Operation used in the given OperationDeclartion, returns
-    /// a Property that returns an instance of the operation by calling the 
+    /// a Property that returns an instance of the operation by calling the
     /// IOperationFactory
     let buildOpsProperties context (operations : QsQualifiedName list): MemberDeclarationSyntax list =
-        let buildOne n =
+        let getModifiers qualifiedName =
+            // Use the right accessibility for the property depending on the accessibility of the callable. Note: In C#,
+            // "private protected" is the intersection of protected and internal.
+            match context.allCallables.[qualifiedName].Modifiers.Access with
+            | DefaultAccess -> [ ``protected`` ]
+            | Internal -> [ ``private``; ``protected`` ]
+        let buildOne qualifiedName =
             /// eg:
             /// protected opType opName { get; }
-            let signature = roslynCallableTypeName context n
-            let name = getOpName context n
-            ``prop`` signature name [ ``protected`` ] 
+            let signature = roslynCallableTypeName context qualifiedName
+            let name = getOpName context qualifiedName
+            let modifiers = getModifiers qualifiedName
+            ``prop`` signature name modifiers
             :> MemberDeclarationSyntax
         operations
         |> List.map buildOne

--- a/src/Simulation/Simulators.Tests/Circuits/CoreOperations.qs
+++ b/src/Simulation/Simulators.Tests/Circuits/CoreOperations.qs
@@ -593,5 +593,24 @@ namespace Microsoft.Quantum.Simulation.Simulators.Tests.Circuits {
 
         AssertEqual(true, (pi > 3.14 and pi < 3.15));
     }
-}
 
+    internal function EmptyInternalFunction() : Unit { }
+
+    internal operation EmptyInternalOperation() : Unit { }
+
+    internal newtype InternalType = Int;
+
+    internal operation MakeInternalType() : InternalType {
+        return InternalType(5);
+    }
+
+    // This is a public operation that uses an internal type inside to test the access modifiers of the generated
+    // operation properties.
+    operation InternalCallablesTest() : Unit {
+        EmptyInternalFunction();
+        EmptyInternalOperation();
+        let x = InternalType(3);
+        let y = MakeInternalType();
+        AssertEqual(15, x! * y!);
+    }
+}

--- a/src/Simulation/Simulators.Tests/CoreTests.cs
+++ b/src/Simulation/Simulators.Tests/CoreTests.cs
@@ -279,5 +279,9 @@ namespace Microsoft.Quantum.Simulation.Simulators.Tests
             }
             Assert.Equal(1, exceptionCount); // check that we cought exception once
         }
+
+        [Fact]
+        public void InternalCallables() =>
+            Helper.RunWithMultipleSimulators(s => Circuits.InternalCallablesTest.Run(s).Wait());
     }
 }

